### PR TITLE
vpn changes 

### PR DESCRIPTION
**Pull request by sdharav**
@njagadish  please review the changes

### CODE REVIEW

**Comment by njagadish** [comment]
@rajendramachani - Could you also review this

File: devstack/lib/nfp [hunk 1]
```diff
@@ -123,10 +123,6 @@ function create_ep_and_nsp {
  --shared True\
  default
 
-    gbp ep-create\
- --external-segments default\
- ext_connect
```
Comment (njagadish) [inline devstack/lib/nfp:128]: Why is this removed?

File: gbpservice/contrib/nfp/configurator/agents/vpn.py [hunk 2]
```diff
@@ -316,10 +314,10 @@ def events_init(sc, drivers):
     Returns: None
     """
     evs = [
-        Event(id='VPNSERVICE_UPDATED',
-              handler=VPNaasEventHandler(sc, drivers)),
-        Event(id='VPN_SYNC',
-              handler=VPNaasEventHandler(sc, drivers))]
+        main.Event(id='VPNSERVICE_UPDATED',
```
Comment (njagadish) [inline gbpservice/contrib/nfp/configurator/agents/vpn.py]: "main" is not meaningful here. Make it nfp_event.

File: gbpservice/nfp/orchestrator/config_drivers/heat_driver.py [hunk 3]
```diff
@@ -820,6 +820,105 @@ def _create_node_config_data(self, auth_token, tenant_id,
                     'description'] = str(common_desc)
 
                 nf_desc = str(firewall_desc)
+        elif service_type == pconst.VPN:
+            # rvpn_l3_policy = self._get_rvpn_l3_policy(auth_token,
+            #    provider, update)
```
Comment (njagadish) [inline gbpservice/nfp/orchestrator/config_drivers/heat_driver.py]: Remove this commented code.

File: gbpservice/nfp/orchestrator/config_drivers/heat_driver.py [hunk 4]
```diff
@@ -820,6 +820,105 @@ def _create_node_config_data(self, auth_token, tenant_id,
...
+            # if rvpn_l3_policy is None:
+            #    return None, None
+            # config_param_values['ClientAddressPoolCidr'] = rvpn_l3_policy[
+            #    'ip_pool']
+            config_param_values['Subnet'] = (
+                consumer_port['fixed_ips'][0]['subnet_id']
+                if consumer_port else None)
+            l2p = self.gbp_client.get_l2_policy(
```
Comment (njagadish) [inline gbpservice/nfp/orchestrator/config_drivers/heat_driver.py:827]: Can we check if l2p and l3p are already retrieved to reuse them. If not, then leave this code as is.

File: gbpservice/nfp/orchestrator/config_drivers/heat_driver.py [hunk 5]
```diff
@@ -820,6 +820,105 @@ def _create_node_config_data(self, auth_token, tenant_id,
...
+                stitching_ptg_id = (
+                    stitching_pts[0]['policy_target_group_id'])
+            else:
+                stitching_ptg_id = consumer['id']
+            self.gbp_client.update_policy_target_group(
+                auth_token, stitching_ptg_id,
+                {'policy_target_group': {
+                    'network_service_policy_id': nsp['id']}})
+            if not base_mode_support:
+                floatingips = self.neutron_client.get_floating_ips(
```
Comment (njagadish) [inline gbpservice/nfp/orchestrator/config_drivers/heat_driver.py:873]: Lets check if all the gets we do here can be avoided by checking if the data is already available in the local variables.

File: devstack/lib/nfp (hunk 1 above)
Comment (sdharav) [inline devstack/lib/nfp:128]: Having one external policy(ep) does not let us create another ep with the consuming rule sets while launching the chain, in such case any consuming rule sets  will not be applied. So it is of no use creating an empty ep here.

File: gbpservice/contrib/nfp/configurator/agents/vpn.py (hunk 2 above)
Comment (sdharav) [inline gbpservice/contrib/nfp/configurator/agents/vpn.py]: addressed.

File: gbpservice/nfp/orchestrator/config_drivers/heat_driver.py (hunk 3 above)
Comment (sdharav) [inline gbpservice/nfp/orchestrator/config_drivers/heat_driver.py]: Removed the unused code.

File: gbpservice/nfp/orchestrator/config_drivers/heat_driver.py (hunk 4 above)
Comment (sdharav) [inline gbpservice/nfp/orchestrator/config_drivers/heat_driver.py:827]: checked that we will not get them before.

File: devstack/lib/nfp (hunk 1 above)
Comment (njagadish) [inline devstack/lib/nfp:128]: So, this is created in the chain create script?

Comment (sdharav) [inline devstack/lib/nfp:128]: yeah, it will be created while launching the chain

File: gbpservice/nfp/orchestrator/config_drivers/heat_driver.py (hunk 5 above)
Comment (YogeshRajmane) [inline gbpservice/nfp/orchestrator/config_drivers/heat_driver.py:873]: Can we add get_floating_ip() in openstack_driver lib file and use that here, so that will reduce overhead of iterating list of floating ips.

File: devstack/lib/nfp (hunk 1 above)
Comment (rajendramachani) [inline devstack/lib/nfp:128]: What about non-VPN services ?
Are you saying that, the ep is created implicitly in any chain during ptg create. Or, in VPN explicitly we are creating ?

Comment (sdharav) [inline devstack/lib/nfp:128]: if not then one should add in the other scripts as well to test for North-South traffic.
